### PR TITLE
#317: Select component

### PR DESCRIPTION
**Pull request by BeeMargarida**
| - | - |
| --- | --- |
| Issue | https://github.com/ripe-tech/ripe-design/issues/317 |
| Dependencies | -- |
| Decisions | Uses [lawnstarter/react-native-picker-select](https://github.com/lawnstarter/react-native-picker-select) package to show a select button and "dropdown". Support for placeholder and width. Ported some utils from sdk. <br><br> Problems with workarounds: <br> • iOS problem mentioned here https://github.com/lawnstarter/react-native-picker-select/issues/344, requires `@react-native-community/picker` as a dependency. <br> • When using a placeholder, the selected value is not highlighted in the dropdown in Android. A fix was made and documented in the code. <br><br> Requires import in storybook and possibly robin-revamp: https://github.com/ripe-tech/ripe-components-react-native-storybook/pull/18 + https://github.com/ripe-tech/ripe-robin-revamp/pull/257  |
| Animated GIF | select-androidselect-ios|

### CODE REVIEW

File: react/components/molecules/select/select.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            iconContainer: {
+                top: "25%",
+                right: 10
+            }
+        };
+    };
+
+    render() {
+        return (
+            <View style={this._style()}>
```
Comment (beneditomauro) [inline react/components/molecules/select/select.js:120]: Id the component

**Review by beneditomauro — COMMENTED** [review]
Id of the component missing

**Review by beneditomauro — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -63,13 +63,15 @@
         "svgo": "^2.3.0"
     },
     "peerDependencies": {
+        "@react-native-community/picker": "^1.8.1",
```
Comment (gcandal) [inline package.json:66]: Can't https://www.npmjs.com/package/@react-native-picker/picker be used instead?

image

Comment (gcandal) [reply package.json:66]: (i've seen your comment on the PR, but still)

File: react/components/molecules/select/select.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+import { IdentifiableMixin, baseStyles, equal } from "../../../util";
+
+import { Icon } from "../../atoms";
+
+export class Select extends mix(PureComponent).with(IdentifiableMixin) {
+    static get propTypes() {
+        return {
+            options: PropTypes.array,
+            value: PropTypes.any,
```
Comment (gcandal) [inline react/components/molecules/select/select.js:15]: How does it behave with empty options?

File: react/components/molecules/select/select.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    onValueChange = value => {
+        this.setState(
+            {
+                valueData: value
+            },
+            () => this.props.onUpdateValue(value)
+        );
+    };
+
+    _icon = () => {
```
Comment (gcandal) [inline react/components/molecules/select/select.js:62]: Aren't we calling this stuff `renderIcon` and keeping it next to `render`? Not sure, just checking.

File: react/components/molecules/select/select.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        return <Icon icon={"chevron-down"} color={"#1b2632"} strokeWidth={2} />;
+    };
+
+    _items = () => {
+        return this.props.options.map(option => ({
+            ...option,
+            key: option.value,
+
+            // workaround for placeholder use overriding
+            // the selected item style
```
Comment (gcandal) [inline react/components/molecules/select/select.js:72]: Can you please reword? Couldn't understand what you mean.

File: package.json (hunk 2 above)
Comment (BeeMargarida) [reply package.json:66]: The library being used is a wrapper around that library that supports the same approach for both ios and android: both showing a button that when clicked opens a picker. The base library (react-native-picker/picker) shows only the picker in iOS and does not support changing the styling of the button (in iOS there is no button). Lanwstarted library allows for a simple API for both platforms, allows overriding the picker button and has the iOS picker behaviour (container opening with the picker inside) out-of-the-box.

File: react/components/molecules/select/select.js (hunk 4 above)
Comment (BeeMargarida) [reply react/components/molecules/select/select.js:62]: Both in the components-vue and react-native, we are using `icon` and `_icon` respectively

File: react/components/molecules/select/select.js (hunk 3 above)
Comment (BeeMargarida) [reply react/components/molecules/select/select.js:15]: Empty select: <br>
| - | - |
| --- | --- |
| iOS | Android |
| imagem | yoyo |

Comment (gcandal) [reply react/components/molecules/select/select.js:15]: Please confirm that ripe-components-vue has the same behavior

File: package.json (hunk 2 above)
Comment (gcandal) [reply package.json:66]: I understand that, my point is: you're including `@react-native-community/picker` which is the deprecated version of `@react-native-picker/picker`. Can't you replaced one with other? (while still using the wrapper `react-native-picker-select`)

File: react/components/molecules/select/select.js (hunk 3 above)
Comment (BeeMargarida) [reply react/components/molecules/select/select.js:15]: Its not the same bahaviour, but porting it is tricky. The problem is in react-native, the only way to do this is to set an option with a `null` value, which triggers a warning. There is a feature request to disable certain options https://github.com/lawnstarter/react-native-picker-select/issues/197, but its not implemented.

| - | - |
| --- | --- |
| components-vue | react-native similar |
| select-components-vue | no-items-select |

**Comment by gcandal** [comment]
Intercept on the value change callback and ignore if "no items"

**Comment by BeeMargarida** [comment]
> Intercept on the value change callback and ignore if "no items"

It works in iOS, but in Android it keeps the `no items` option selected, even after forcing the value selected to null:
| - | - |
| --- | --- |
| iOS | Android |
| select-no-items-ios | select-no-items-android|

I know its different from `components-vue`, but perhaps showing in the placeholder that the select is empty (by changing the placeholder text to `No items` or appending to the placeholder something like `(empty)`) and disabling the button. Or leaving as is, with the placeholder as the only option.

**Review by gcandal — APPROVED** [review]
New dependencies:
- https://www.npmjs.com/package/@react-native-picker/picker
- https://www.npmjs.com/package/react-native-picker-select

**Review by joamag — APPROVED** [review]

**Pull request merged**